org.eclipse.scout.rt.oauth2/src/main/java/org/eclipse/scout/rt/oauth2/OAuth2Helper.java

```diff
@@ -0,0 +1,118 @@
+import org.eclipse.scout.rt.platform.util.date.IDateProvider;
+
+import com.github.scribejava.core.builder.ServiceBuilder;
+import com.github.scribejava.core.builder.api.DefaultApi20;
+import com.github.scribejava.core.model.OAuth2AccessToken;
+import com.github.scribejava.core.oauth.OAuth20Service;
+
+@ApplicationScoped
+public class OAuth2Helper {
+
+  protected final ConcurrentHashMap<OAuth2Config, TokenEntry> m_tokenCache = new ConcurrentHashMap<>();
+  protected final IDateProvider m_dateProvider = BEANS.get(IDateProvider.class);
+  protected long m_latestEviction = m_dateProvider.currentUTCMillis();
+  // Clean map after 30 minutes (or longer, if #getToken() isn't called)
+  protected final static long EVICTION_INTERVALL = 30 * 60 * 1000;
+
+  /**
+   * Supports the OAuth2 client credentials flow.
+   *
+   * @return OAuth2 access token
+   */
+  public String getToken(OAuth2Config oAuth2Config) {
+    try {
+      TokenEntry tokenEntry = null;
+      tokenEntry = m_tokenCache.get(oAuth2Config);
+      if (tokenEntry == null
+          || tokenEntry.getAccessToken() == null
+          || tokenEntry.getTokenReceivedTime() == null
+          // If expired or when expiring within the next minute: Get a new token
+          || (tokenEntry.getTokenReceivedTime() + tokenEntry.getAccessToken().getExpiresIn() * 1000 - 60000) < m_dateProvider.currentUTCMillis()) {
+        OAuth2AccessToken token;
+        try (OAuth20Service service = new ServiceBuilder(oAuth2Config.getClientId())
+            .apiSecret(oAuth2Config.getClientSecret())
+            .defaultScope(oAuth2Config.getScope())
+            .build(new DefaultApi20() {
+
+              @Override
+              public String getAccessTokenEndpoint() {
+                return oAuth2Config.getTokenEndpoint();
+              }
+
+              @Override
+              protected String getAuthorizationBaseUrl() {
+                return oAuth2Config.getAuthorizationEndpoint();
+              }
+            })) {
+          token = service.getAccessTokenClientCredentialsGrant();
+
+          tokenEntry = new TokenEntry(token, m_dateProvider.currentUTCMillis());
+          m_tokenCache.put(oAuth2Config, tokenEntry);
+        }
+        catch (IOException | InterruptedException | ExecutionException e) {
+          throw new ProcessingException("Exception while retrieving OAuth2 access token.", e);
+        }
+      }
+      if (tokenEntry.getAccessToken() != null) {
+        return tokenEntry.getAccessToken().getAccessToken();
+      }
+      return null;
+    }
+    finally {
+      evictExpiredEntries();
+    }
+  }
+
+  /**
+   * Evict all expired entries when at least EVICTION_INTERVALL milliseconds have elapsed since last eviction
+   */
+  protected void evictExpiredEntries() {
+    if (m_latestEviction + EVICTION_INTERVALL > m_dateProvider.currentUTCMillis()) {
+      for (Entry<OAuth2Config, TokenEntry> entry : m_tokenCache.entrySet()) {
```

Comment:
Use Iteratore here? Or is it okay to remove from entry set while iterating? The JavaDoc might lead to this conclusion, not sure though.